controller/kubernetes_pod_controller.go

```diff
@@ -0,0 +1,289 @@
+
+	queue workqueue.RateLimitingInterface
+
+	//// for unit test
+	//nowHandler func() string
+}
+
+func NewKubernetesPodController(
+	ds *datastore.DataStore,
+	scheme *runtime.Scheme,
+	kubePodInformer coreinformers.PodInformer,
+	kubePersistentVolumeInformer coreinformers.PersistentVolumeInformer,
+	kubePersistentVolumeClaimInformer coreinformers.PersistentVolumeClaimInformer,
+	kubeClient clientset.Interface,
+	controllerID string) *KubernetesPodController {
+
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events(""),
+	})
+
+	kc := &KubernetesPodController{
+		controllerID: controllerID,
+
+		ds: ds,
+
+		kubeClient:    kubeClient,
+		eventRecorder: eventBroadcaster.NewRecorder(scheme, v1.EventSource{Component: controllerAgentName}),
+
+		pLister:   kubePodInformer.Lister(),
+		pvLister:  kubePersistentVolumeInformer.Lister(),
+		pvcLister: kubePersistentVolumeClaimInformer.Lister(),
+
+		pStoreSynced:   kubePodInformer.Informer().HasSynced,
+		pvStoreSynced:  kubePersistentVolumeInformer.Informer().HasSynced,
+		pvcStoreSynced: kubePersistentVolumeClaimInformer.Informer().HasSynced,
+
+		queue: workqueue.NewNamedRateLimitingQueue(EnhancedDefaultControllerRateLimiter(), "longhorn-kubernetes-pod"),
+	}
+
+	kubePodInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			curPod := cur.(*v1.Pod)
+			kc.enqueuePodChange(curPod)
+		},
+		DeleteFunc: func(obj interface{}) {
+			pod := obj.(*v1.Pod)
+			kc.enqueuePodChange(pod)
+		},
+	})
+
+	return kc
+}
+
+func (kc *KubernetesPodController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer kc.queue.ShutDown()
+
+	logrus.Infof("Start %v", controllerAgentName)
+	defer logrus.Infof("Shutting down %v", controllerAgentName)
+
+	if !controller.WaitForCacheSync(controllerAgentName, stopCh, kc.pStoreSynced, kc.pvStoreSynced, kc.pvcStoreSynced) {
+		return
+	}
+	for i := 0; i < workers; i++ {
+		go wait.Until(kc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (kc *KubernetesPodController) worker() {
+	for kc.processNextWorkItem() {
+	}
+}
+
+func (kc *KubernetesPodController) processNextWorkItem() bool {
+	key, quit := kc.queue.Get()
+	if quit {
+		return false
+	}
+	defer kc.queue.Done(key)
+	err := kc.syncHandler(key.(string))
+	kc.handleErr(err, key)
+	return true
+}
+
+func (kc *KubernetesPodController) handleErr(err error, key interface{}) {
+	if err == nil {
+		kc.queue.Forget(key)
+		return
+	}
+
+	if kc.queue.NumRequeues(key) < maxRetries {
+		logrus.Warnf("%v: Error syncing Longhorn kubernetes pod %v: %v", controllerAgentName, key, err)
```

Comment:
Would be good to keep the message consistent with the other controllers, suggest dropping the `controllerAgentName`:
```
logrus.Warnf("Error syncing Longhorn kubernetes pod %v: %v", key, err)
```

Reply:
Thank you @c3y1huang for helping!
Yeah, adding `controllerAgentName ` is inconsistent with other controllers. However, one problem I had when reading log messages is that it is hard to tell which where the log messages came from. 

@joshimoo How do you think about this?

Reply:
Thanks @c3y1huang . In fact, I prefer to have the controller name in the controller log, so we know which controller is printing out the log. Unfortunately, we didn't apply this practice across different controllers, and that's something we can improve.

Also, I am wondering if there is an easier way to do it, rather than manually set the controllerName every time.

Reply:
Yes there is, if we refactor to a structured logging approach, (https://godoc.org/gopkg.in/Sirupsen/logrus.v0#FieldLogger)

We can init a logger at controller creation with the generic fields at controller creation.
Then also init a specific logger for the resource type with the relevant fields.

An example would look like this:
baseLogger -> controllerLogger(baseLogger, agentName) -> podLogger(controllerLogger, podName, ...)

@yasker I can push a quick refactor of this, if we agree?

Reply:
- [x] created an epic https://github.com/longhorn/longhorn/issues/1724
- [x] created tasks for each controller
- [x] implemented support for structured instance based logging in longhorn/longhorn-manager#661
- [x] refactored engine-controller to use instance based structured logging in longhorn/longhorn-manager#662

I marked the individual controller issues with `good first issue`, since these are excellent issues for new contributors.
- low complexity
- low domain knowledge (kubernetes controller, logging)
- independent (each controller on it's own)
- exposure to a small parts of the longhorn code flow (reading the controller code)

So if someone asks how they can contribute feel free to point them to one of these issues.